database/download.go

```diff
@@ -52,48 +61,84 @@ func (db *DB) DownloadCreate(ctx context.Context, user User, skylink Skylink) (*
 }
 
 // DownloadsBySkylink fetches all downloads of this skylink
-func (db *DB) DownloadsBySkylink(ctx context.Context, skylink Skylink, offset, limit int) ([]Download, error) {
+func (db *DB) DownloadsBySkylink(ctx context.Context, skylink Skylink, offset, limit int) ([]DownloadResponseDTO, error) {
 	if skylink.ID.IsZero() {
 		return nil, errors.New("invalid skylink")
 	}
-	filter := bson.D{{"skylink_id", skylink.ID}}
-	opts := options.FindOptions{}
-	if offset > 0 {
-		opts.SetSkip(int64(offset))
-	}
-	if limit > 0 {
-		opts.SetLimit(int64(limit))
-	}
-	c, err := db.staticDownloads.Find(ctx, filter, &opts)
-	if err != nil {
-		return nil, err
-	}
-	downloads := make([]Download, 0)
-	err = c.All(ctx, &downloads)
-	if err != nil {
-		return nil, err
-	}
-	return downloads, nil
+	matchStage := bson.D{{"$match", bson.D{{"skylink_id", skylink.ID}}}}
+	return db.downloadsBy(ctx, matchStage, offset, limit)
 }
 
 // DownloadsByUser fetches all downloads by this user
-func (db *DB) DownloadsByUser(ctx context.Context, user User, offset, limit int) ([]Download, error) {
+func (db *DB) DownloadsByUser(ctx context.Context, user User, offset, limit int) ([]DownloadResponseDTO, error) {
 	if user.ID.IsZero() {
 		return nil, errors.New("invalid user")
 	}
-	filter := bson.D{{"user_id", user.ID}}
-	opts := options.FindOptions{}
-	if offset > 0 {
-		opts.SetSkip(int64(offset))
+	matchStage := bson.D{{"$match", bson.D{{"user_id", user.ID}}}}
+	return db.downloadsBy(ctx, matchStage, offset, limit)
+}
+
+// downloadsBy is a helper function that allows us to fetch a list of downloads,
+// filtered by an arbitrary match criteria.
+//
+// The Mongo query this method executes is
+//	db.downloads.aggregate([
+//		{ $match: { "user_id": ObjectId("5fda32ef6e0aba5d16c0d550") }},
+//		{ $skip: 1 },
+//		{ $limit: 5 },
+//		{ $lookup: {
+//				from: "skylinks",
+//				localField: "skylink_id",  // field in the downloads collection
+//				foreignField: "_id",	   // field in the skylinks collection
+//				as: "fromSkylinks"
+//		  }
+//		},
+//		{ $replaceRoot: { newRoot: { $mergeObjects: [ { $arrayElemAt: [ "$fromSkylinks", 0 ] }, "$$ROOT" ] } } },
+//		{ $project: { fromSkylinks: 0 } },
+//		{ $addFields: { string_id: { $toString: "$_id" } } }
+//	])
+//
+// This query will get all downloads by the current user, skip $skip of them
+// and then fetch $limit of them, allowing us to paginate. It will then
+// join with the `skylinks` collection in order to fetch some additional
+// data about each download. The last line converts the [12]byte `_id` to hex,
+// so we can easily handle it in JSON.
```

Comment:
The driver should handle this for you actually, pretty sure it would call `toHexString()` on the `ObjectId`. But I don't have that much against the `$toString` you did here either.

Reply:
Oh, you're right! I was trying before with `int64` instead of `time.Time` and it didn't work but it does work with `time.Time`, so I'm getting rid of the `$toString`. Thanks!